types/decimal.go

```diff
@@ -222,123 +226,192 @@ func (d Dec) BigInt() *big.Int {
 	return copy.Set(d.i)
 }
 
+func (d Dec) ImmutOp(op func(Dec, Dec) Dec, d2 Dec) Dec {
```

Comment:
I'm suspicious whether this adds overhead or not, but perhaps we should benchmark that. I don't know how good golang is at inlining, but I also don't think code duplication here is that bad.